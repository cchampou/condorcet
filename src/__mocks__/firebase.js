import { when } from "jest-when";

import { MOCK_POLL_ID, MOCK_USER_ID, MOCK_POLL } from "../utils/test";

const doc = jest.fn();
const votesSubDoc = jest.fn();
const resultsSubDoc = jest.fn();
const subCollection = jest.fn();

when(votesSubDoc)
  .calledWith(MOCK_USER_ID)
  .mockReturnValue({
    set: jest.fn(() => Promise.resolve())
  });

when(subCollection)
  .calledWith("votes")
  .mockReturnValue({
    doc: votesSubDoc
  });

when(resultsSubDoc)
  .calledWith("condorcet")
  .mockReturnValue({
    get: () => ({
      data: () => ({
        matrix: {
          zorro: {
            luffy: 1,
            sanji: 1
          },
          luffy: {
            zorro: 0,
            sanji: 1
          },
          sanji: {
            zorro: 0,
            luffy: 0
          }
        },
        ranking: [
          {
            rank: 1,
            value: "Zorro",
            slug: "zorro",
            wins: 1,
            equalities: 1
          },
          {
            rank: 2,
            value: "Luffy",
            slug: "luffy",
            wins: 1,
            equalities: 1
          },
          {
            rank: 3,
            value: "Sanji",
            slug: "sanji",
            wins: 0,
            equalities: 0
          }
        ]
      })
    })
  });

when(resultsSubDoc)
  .calledWith("uninominal")
  .mockReturnValue({
    get: () => ({
      data: () => ({
        ranking: [{ value: "luffy", rank: 1 }]
      })
    })
  });

when(subCollection)
  .calledWith("results")
  .mockReturnValue({
    doc: resultsSubDoc
  });

when(doc)
  .calledWith(MOCK_POLL_ID)
  .mockReturnValue({
    get: jest.fn(() => ({
      data: jest.fn(() => ({ ...MOCK_POLL, collection: subCollection }))
    })),
    collection: subCollection,
    set: jest.fn()
  });

module.exports = {
  auth: {
    signOut: jest.fn(() => Promise.resolve()),
    signInWithPopup: jest.fn(() => Promise.resolve({ user: "Smaug" })),
    useDeviceLanguage: jest.fn(),
    setPersistence: jest.fn(),
    onAuthStateChanged: jest.fn(() => ({ id: MOCK_USER_ID })),
    GoogleAuthProvider: jest.fn(),
    GithubAuthProvider: jest.fn(),
    FacebookAuthProvider: jest.fn(),
    Auth: {
      Persistence: { LOCAL: "LOCAL" }
    }
  },
  db: {
    collection: jest.fn(() => ({
      where: jest.fn(() => ({
        get: () => [
          {
            data: () => {}
          }
        ],
        orderBy: jest.fn(() => ({
          onSnapshot: jest.fn(() => [MOCK_POLL])
        }))
      })),
      add: jest.fn(() => Promise.resolve()),
      doc
    }))
  },
  firestore: {
    FieldValue: {
      serverTimestamp() {
        return {};
      }
    }
  }
};
