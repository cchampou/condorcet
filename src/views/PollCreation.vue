<template>
  <div>
    <input type="text" label="Nom" v-model="question" />

    <textarea
      name="input-7-1"
      label="Description"
      v-model="description"
    ></textarea>

    <input
      type="text"
      label="Ajouter une réponse"
      v-model="answer"
      @keypress.enter="addAnswer"
    />
    <button color="info" @click="addAnswer">Ajouter</button>

    <div>
      <div class="panel-block" v-for="answer in answers" :key="answer.slug">
        {{ answer.value }}
      </div>
    </div>

    <button color="info" @click="createPoll">Créer</button>
  </div>
</template>

<script>
import { mapState, mapActions } from "vuex";
import randomString from "random-string";
import slugify from "slugify";
import { db } from "../firebase";

export default {
  name: "PollCreation",
  data() {
    return {
      question: "",
      description: "",
      answer: "",
      answers: []
    };
  },
  computed: {
    ...mapState({
      userId: state => state.user.id
    })
  },
  methods: {
    ...mapActions("app", ["addNotification"]),
    async createPoll() {
      if (this.question === "" || this.answers.length < 2) {
        this.addNotification({
          text: "Veuillez renseigner l'ensemble des informations requises",
          status: "error"
        });
        return;
      }

      try {
        let users = [this.userId];

        await db.collection("polls").add({
          question: this.question,
          description: this.description,
          answers: this.answers,
          owner: this.userId,
          isActive: true,
          token: randomString({ length: 20 }),
          users,
          condorcet: null,
          uninominal: null
        });

        this.addNotification({
          text: "Scrutin créé mon pote",
          status: "success"
        });

        this.name = "";
        this.description = "";
        this.answer = "";
        this.answers = [];
      } catch (error) {
        this.addNotification({
          text: `Il y a eu un problème lors de la création du scrutin`,
          status: "error"
        });
      }
    },
    addAnswer() {
      if (this.answers.find(answer => answer.value === this.answer)) {
        this.addNotification({
          text: "Cette réponse existe déjà",
          status: "warning"
        });

        return;
      }

      this.answers = [
        ...this.answers,
        {
          value: this.answer,
          slug: slugify(this.answer, { lower: true, replacement: "-" })
        }
      ];
      this.answer = "";
    }
  }
};
</script>

<style></style>
