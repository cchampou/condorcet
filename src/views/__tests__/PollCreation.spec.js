import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import PollCreation from "../PollCreation.vue";
import firebase from "../../firebase";

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock("../../firebase.js", () => ({
  db: {
    collection: jest.fn(() => ({
      add: jest.fn(() => Promise.resolve())
    }))
  }
}));

describe("PollCreation.vue", () => {
  let wrapper;
  let appActions;

  beforeEach(() => {
    jest.clearAllMocks();
    appActions = { addNotification: jest.fn() };

    const store = new Vuex.Store({
      modules: {
        app: { namespaced: true, actions: appActions },
        user: {
          namespaced: true
        }
      }
    });

    wrapper = shallowMount(PollCreation, {
      localVue,
      store,
      computed: {
        userId: () => "azfnez893jfz"
      }
    });
  });

  it("should render", () => {
    expect(wrapper.vm.userId).toBe("azfnez893jfz");
    expect(wrapper).toMatchSnapshot();
  });

  it("should not add answer to the poll", () => {
    wrapper.vm.addAnswer();

    expect(appActions.addNotification).toHaveBeenCalled();
  });

  it("should add answer to the poll", () => {
    const answerInput = wrapper.find("#answer");
    const addAnwserButton = wrapper.find("button");

    answerInput.setValue("Les pommes");
    answerInput.trigger("change");

    expect(wrapper.vm.answer).toBe("Les pommes");

    addAnwserButton.trigger("click");

    expect(wrapper.vm.answer).toBe("");
    expect(wrapper.vm.answers).toEqual([
      {
        slug: "les-pommes",
        value: "Les pommes"
      }
    ]);
  });

  it("should not create poll and notify user that there is an error if question or answer are not set", () => {
    wrapper.vm.createPoll();

    expect(appActions.addNotification).toHaveBeenCalled();
  });

  it("should create the poll and notify user that there is an error", () => {
    wrapper.setData({
      question: "Qui est la plus belle ?",
      answers: ["Megan Fox", "Scarlett", "Natalie Portman"]
    });

    wrapper.vm.createPoll();

    expect(firebase.db.collection).toHaveBeenCalledWith("polls");
  });
});
