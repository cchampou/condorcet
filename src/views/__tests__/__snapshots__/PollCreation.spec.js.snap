// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollCreation.vue should render 1`] = `
<div>
  <form autocomplete="off" class="w-full">
    <div class="flex flex-wrap -mx-3 mb-6">
      <div class="w-full md:w-1/2 px-3 mb-6 md:mb-0"><label for="grid-first-name" class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2">
          Question *
        </label> <input id="grid-first-name" type="text" class="appearance-none block w-full bg-gray-200 text-gray-700 border border-red-500 rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white">
        <div class="w-full md:w-1/2 px-3"><textarea class="appearance-none block w-full bg-gray-200 text-gray-700 border border-red-500 rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"></textarea></div>
      </div>
      <div class="w-full md:w-1/2 px-3"><label for="answer" class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2">
          Ajouter des réponses *
        </label> <input id="answer" type="text" class="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"> <button class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
          Ajouter
        </button>
        <div>
          <ul></ul>
        </div>
      </div>
    </div> <button class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
      Créer
    </button>
  </form>
</div>
`;
